src/Microsoft.Azure.WebJobs.Extensions.Kafka/Listeners/KafkaMetricsProvider.cs

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Confluent.Kafka;
+using Microsoft.Azure.WebJobs.Host.Scale;
+using Microsoft.Extensions.Logging;
+using System;
+using System.CodeDom.Compiler;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Extensions.Kafka;
+using static Confluent.Kafka.ConfigPropertyNames;
+
+
+namespace Microsoft.Azure.WebJobs.Extensions.Kafka
+{
+    public class KafkaMetricsProvider<TKey, TValue>
+    {
+        private readonly string topicName;
+        private readonly AdminClientConfig adminClientConfig;
+        private readonly IConsumer<TKey, TValue> consumer;
+        private readonly ILogger logger;
+        protected Lazy<List<TopicPartition>> topicPartitions;
+
+        virtual protected internal KafkaTriggerMetrics LastCalculatedMetrics { get; set; }
+
+        public KafkaMetricsProvider(string topicName, AdminClientConfig adminClientConfig, IConsumer<TKey, TValue> consumer, ILogger logger)
+        {
+            this.topicName = topicName;
+            this.adminClientConfig = adminClientConfig;
+            this.logger = logger;
+            this.consumer = consumer;
+            this.topicPartitions = new Lazy<List<TopicPartition>>(LoadTopicPartitions);
+            this.LastCalculatedMetrics = null;
+        }
+
+        public virtual Task<KafkaTriggerMetrics> GetMetricsAsync()
+        {
+            var allPartitions = topicPartitions.Value;
+            if (allPartitions == null)
+            {
+                return Task.FromResult(new KafkaTriggerMetrics(0L, 0));
+            }
+
+            var operationTimeout = TimeSpan.FromSeconds(5);
+
+            long totalLag = 0;
+            try
+            {
+                totalLag = GetTotalLag(allPartitions, operationTimeout);
+            }
+            catch (Exception ex)
+            {
+                logger.LogError(ex, $"Failed to load retrieve lag from topic '{this.topicName}'");
+            }
+            int paritionCount = allPartitions.Count;
+
+            var metrics = new KafkaTriggerMetrics(totalLag, paritionCount);
+
+            // Storing the metrics along with TimeStamp.
+            this.LastCalculatedMetrics = metrics;
+
+            return Task.FromResult(metrics);
+        }
+
+        protected virtual List<TopicPartition> LoadTopicPartitions()
```

Comment:
Why is it virtual? For tests mocking?

Reply:
Yes, it is virtual for testing purposes although rather than mocking, it is to be able to override it for unit tests.